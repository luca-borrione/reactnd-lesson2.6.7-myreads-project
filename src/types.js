
import PropTypes from 'prop-types'

/**
 * @typedef {Object} TBook
 * @description
 * Custom type to define a book internally.
 * It's filled with properties coming from the
 * book literal object fetched using the BooksAPI
 * @property {string[]} [authors] - List of the authors names
 * @property {string} id - The book unique id
 * @property {string} [shelf] - The shelf unique name
 * @property {string} [thumbnail] - URL for the cover preview image
 * @property {string} title - The book's title
 */

export const TBook = {
	authors: PropTypes.arrayOf(PropTypes.string.isRequired),
	id: PropTypes.string.isRequired,
	shelf: PropTypes.string,
	thumbnail: PropTypes.string,
	title: PropTypes.string.isRequired
};

