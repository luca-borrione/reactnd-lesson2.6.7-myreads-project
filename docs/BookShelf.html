<!DOCTYPE html>
<html lang="en">
<head>
    
    <meta charset="utf-8">
    <title>BookShelf - Documentation</title>
    
    
    <script src="scripts/prettify/prettify.js"></script>
    <script src="scripts/prettify/lang-css.js"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc.css">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>
<body>

<input type="checkbox" id="nav-trigger" class="nav-trigger" />
<label for="nav-trigger" class="navicon-button x">
  <div class="navicon"></div>
</label>

<label for="nav-trigger" class="overlay"></label>

<nav>
    
    <h2><a href="index.html">Home</a></h2><h3>Classes</h3><ul><li><a href="App.html">App</a><ul class='methods'><li data-type='method'><a href="App.html#fetchAllBooks">fetchAllBooks</a></li><li data-type='method'><a href="App.html#getBookShelf">getBookShelf</a></li><li data-type='method'><a href="App.html#render">render</a></li><li data-type='method'><a href="App.html#updateBookShelf">updateBookShelf</a></li></ul></li><li><a href="Book.html">Book</a><ul class='methods'><li data-type='method'><a href="Book.html#render">render</a></li></ul></li><li><a href="BooksGrid.html">BooksGrid</a><ul class='methods'><li data-type='method'><a href="BooksGrid.html#render">render</a></li></ul></li><li><a href="BookShelf.html">BookShelf</a><ul class='methods'><li data-type='method'><a href="BookShelf.html#render">render</a></li></ul></li><li><a href="BookShelfChanger.html">BookShelfChanger</a><ul class='methods'><li data-type='method'><a href="BookShelfChanger.html#onSelectChange">onSelectChange</a></li><li data-type='method'><a href="BookShelfChanger.html#render">render</a></li></ul></li><li><a href="BooksList.html">BooksList</a><ul class='methods'><li data-type='method'><a href="BooksList.html#render">render</a></li></ul></li><li><a href="MyReads.html">MyReads</a><ul class='methods'><li data-type='method'><a href="MyReads.html#render">render</a></li></ul></li><li><a href="SearchPage.html">SearchPage</a><ul class='methods'><li data-type='method'><a href="SearchPage.html#render">render</a></li></ul></li></ul><h3>Modules</h3><ul><li><a href="module-ArrayUtils.html">ArrayUtils</a><ul class='methods'><li data-type='method'><a href="module-ArrayUtils.html#~diff">diff</a></li><li data-type='method'><a href="module-ArrayUtils.html#~diffByKey">diffByKey</a></li></ul></li><li><a href="module-Constants.html">Constants</a></li><li><a href="module-StringUtils.html">StringUtils</a><ul class='methods'><li data-type='method'><a href="module-StringUtils.html#~makeTitle">makeTitle</a></li><li data-type='method'><a href="module-StringUtils.html#~ucfirst">ucfirst</a></li><li data-type='method'><a href="module-StringUtils.html#~ucwords">ucwords</a></li></ul></li></ul><h3>Global</h3><ul><li><a href="global.html#TShelfKey">TShelfKey</a></li></ul>
</nav>

<div id="main">
    
    <h1 class="page-title">BookShelf</h1>
    

    




<section>

<header>
    
        <h2>
        BookShelf
        </h2>
        
            <div class="class-description"><p>Prints the shelf title and shows the list of all the books in the shelf</p></div>
        
    
</header>

<article>
    

    
        <h3 class="subsection-title">Extends</h3>

        


    <ul>
        <li>React.Component</li>
    </ul>


    

    

    

     

    

    
        <h3 class="subsection-title">Members</h3>

        
            
<h4 class="name" id="propTypes"><span class="type-signature"></span>propTypes<span class="type-signature"></span></h4>





<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    
</dl>



    <h5 class="subsection-title">Properties:</h5>

    

<table class="props">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>propTypes</code></td>
            

            <td class="type">
            
                
<span class="param-type">Object</span>


            
            </td>

            

            

            <td class="description last"><p>Intended types passed to the component</p>
                <h6>Properties</h6>

<table class="props">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>books</code></td>
            

            <td class="type">
            
                
<span class="param-type">Array.&lt;<a href="global.html#TBook">TBook</a>></span>


            
            </td>

            

            

            <td class="description last"><p>List of all the books in the shelf</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>updateBookShelf</code></td>
            

            <td class="type">
            
                
<span class="param-type">function</span>


            
            </td>

            

            

            <td class="description last"><p>Method of the <a href="App.html#updateBookShelf">App</a> component for updating the list of books
    in the shelves</p></td>
        </tr>

    
    </tbody>
</table>

            </td>
        </tr>

    
    </tbody>
</table>













        
    

    
        <h3 class="subsection-title">Methods</h3>

        
            

    

    <h4 class="name" id="render"><span class="type-signature"></span>render<span class="signature">()</span><span class="type-signature"> &rarr; {ReactElement}</span></h4>

    




<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    
</dl>





<div class="description">
    <ul>
<li>Prints the shelf's title</li>
<li>Shows the list of books in the shelf by rendering the <a href="BooksGrid.html">BooksGrid</a> component</li>
</ul>
</div>





















<h5>Throws:</h5>

        

<dl>
    <dt>
        <div class="param-desc">
        <p>if the books passed to the component are contained in multiple shelves an exception will be thrown</p>
        </div>
    </dt>
    <dd></dd>
    <dt>
        <dl class="param-type">
            <dt>
                Type
            </dt>
            <dd>
                
<span class="param-type">Error</span>


            </dd>
        </dl>
    </dt>
    <dd></dd>
</dl>


    


<h5>Returns:</h5>

        


<dl class="param-type">
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">ReactElement</span>


    </dd>
</dl>

    


        
    

    

    
</article>

</section>




    
    
</div>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.5.5</a> on Tue Nov 20 2018 07:01:22 GMT+0000 (GMT) using the <a href="https://github.com/clenemt/docdash">docdash</a> theme.
</footer>

<script>prettyPrint();</script>
<script src="scripts/linenumber.js"></script>


</body>
</html>