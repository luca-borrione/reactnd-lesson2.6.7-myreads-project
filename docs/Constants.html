<!DOCTYPE html>
<html lang="en">
<head>
    
    <meta charset="utf-8">
    <title>Constants - Documentation</title>
    
    
    <script src="scripts/prettify/prettify.js"></script>
    <script src="scripts/prettify/lang-css.js"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc.css">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>
<body>

<input type="checkbox" id="nav-trigger" class="nav-trigger" />
<label for="nav-trigger" class="navicon-button x">
  <div class="navicon"></div>
</label>

<label for="nav-trigger" class="overlay"></label>

<nav>
    
    <h2><a href="index.html">Home</a></h2><h3>Classes</h3><ul><li><a href="App.html">App</a><ul class='methods'><li data-type='method'><a href="App.html#addBookToShelf">addBookToShelf</a></li><li data-type='method'><a href="App.html#fetchAllBooks">fetchAllBooks</a></li><li data-type='method'><a href="App.html#getBookShelf">getBookShelf</a></li><li data-type='method'><a href="App.html#moveBookToShelf">moveBookToShelf</a></li><li data-type='method'><a href="App.html#removeBook">removeBook</a></li><li data-type='method'><a href="App.html#render">render</a></li><li data-type='method'><a href="App.html#updateBookShelf">updateBookShelf</a></li></ul></li><li><a href="Book.html">Book</a><ul class='methods'><li data-type='method'><a href="Book.html#onShelfChange">onShelfChange</a></li><li data-type='method'><a href="Book.html#render">render</a></li></ul></li><li><a href="BooksGrid.html">BooksGrid</a><ul class='methods'><li data-type='method'><a href="BooksGrid.html#render">render</a></li></ul></li><li><a href="BookShelf.html">BookShelf</a><ul class='methods'><li data-type='method'><a href="BookShelf.html#render">render</a></li></ul></li><li><a href="BookShelfChanger.html">BookShelfChanger</a><ul class='methods'><li data-type='method'><a href="BookShelfChanger.html#render">render</a></li><li data-type='method'><a href="BookShelfChanger.html#selectionChanged">selectionChanged</a></li></ul></li><li><a href="BooksList.html">BooksList</a><ul class='methods'><li data-type='method'><a href="BooksList.html#render">render</a></li></ul></li><li><a href="Search.html">SearchPage</a><ul class='methods'><li data-type='method'><a href="Search.html#clearBooksGrid">clearBooksGrid</a></li><li data-type='method'><a href="Search.html#onSearchTyping">onSearchTyping</a></li><li data-type='method'><a href="Search.html#render">render</a></li><li data-type='method'><a href="Search.html#search">search</a></li></ul></li></ul><h3>Namespaces</h3><ul><li><a href="Constants.html">Constants</a></li><li><a href="StringUtils.html">StringUtils</a><ul class='methods'><li data-type='method'><a href="StringUtils.html#.makeTitle">makeTitle</a></li><li data-type='method'><a href="StringUtils.html#.ucFirst">ucFirst</a></li><li data-type='method'><a href="StringUtils.html#.ucWords">ucWords</a></li></ul></li></ul><h3>Global</h3><ul><li><a href="global.html#TShelfKey">TShelfKey</a></li></ul>
</nav>

<div id="main">
    
    <h1 class="page-title">Constants</h1>
    

    




<section>

<header>
    
        <h2>
        Constants
        </h2>
        
    
</header>

<article>
    
        <div class="container-overview">
        
            

<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    
</dl>




            

            
        
        </div>
    

    

    

    

     

    

    
        <h3 class="subsection-title">Members</h3>

        
            
<h4 class="name" id=".SHELF_TITLE"><span class="type-signature">(static, constant) </span>SHELF_TITLE<span class="type-signature"> :<a href="global.html#TShelfTitle">TShelfTitle</a></span></h4>





<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    

    
</dl>







    <h5>Type:</h5>
    <ul>
        <li>
            
<span class="param-type"><a href="global.html#TShelfTitle">TShelfTitle</a></span>


        </li>
    </ul>






        
    

    

    

    
</article>

</section>




    
    
</div>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.5.5</a> on Sat Nov 03 2018 19:09:19 GMT+0000 (GMT) using the <a href="https://github.com/clenemt/docdash">docdash</a> theme.
</footer>

<script>prettyPrint();</script>
<script src="scripts/linenumber.js"></script>


</body>
</html>