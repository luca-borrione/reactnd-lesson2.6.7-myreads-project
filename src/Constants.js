
import { TShelfKey } from './types'

/**
 * @namespace Constants
 */

const Constants = {};


/**
 * @typedef {Object} TShelfTitle
 * @property {TShelfKey} TShelfKey.CURRENTLY_READING - Currently Reading
 * @property {TShelfKey} TShelfKey.WANT_TO_READ - Want to Read
 * @property {TShelfKey} TShelfKey.READ - Read
 * @property {TShelfKey} TShelfKey.NONE - None
 */
/**
 * @memberof Constants
 * @name SHELF_TITLE
 * @type {TShelfTitle}
 * @constant
 */
export const SHELF_TITLE = Constants.SHELF_TITLE = Object.freeze({
	[TShelfKey.CURRENTLY_READING]: 'Currently Reading',
	[TShelfKey.WANT_TO_READ]: 'Want to Read',
	[TShelfKey.READ]: 'Read',
	[TShelfKey.NONE]: 'None',
});


export default Constants
