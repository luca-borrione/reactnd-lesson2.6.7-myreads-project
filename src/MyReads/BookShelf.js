import React from 'react';
import BooksGrid from '../shared/BooksGrid';
import PropTypes from 'prop-types';
import { TBook } from '../shared/types';
import { ERROR, SHELF_TITLE } from '../shared/constants';

/**
 * @module
 * @name BookShelf
 * @extends React.Component
 * @description
 * Prints the shelf title and shows the list of all the books in the shelf
 */
class BookShelf extends React.Component {

	/**
	 * @member
	 * @name propTypes
	 * @description Intended types of the props passed to the component
	 * @property {TBook[]} books - List of all the books in the shelf
	 * @property {function} updateBookShelf - see [updateBookShelf]{@link module:App~updateBookShelf}
	 * @static
	 */
	static propTypes = {
		books: PropTypes.arrayOf(
			PropTypes.shape(TBook).isRequired
		).isRequired,
		updateBookShelf: PropTypes.func.isRequired
	};


	/**
	 * @method
	 * @name render
	 * @description
	 * - Prints the shelf's title
	 * - Shows the list of books in the shelf by rendering the [BooksGrid]{@link module:BooksGrid} component
	 * @returns {ReactElement}
	 * @throws {Error} if the books passed to the component are contained in multiple shelves an exception will be thrown
	 */
	render() {
		const { books, updateBookShelf } = this.props;

		const shelfKeys = [
			...new Set(books.map( book => book.shelf ))
		];

		if (shelfKeys.length > 1) {
			throw new Error(ERROR.BOOKS_IN_MULTI_SHELVES, shelfKeys);
		}
		const shelfTitle = SHELF_TITLE[shelfKeys[0]];

		return (
			<div className="bookshelf">
				<h2 className="bookshelf-title">{shelfTitle}</h2>
				<div className="bookshelf-books">

					<BooksGrid
						books={books}
						updateBookShelf={updateBookShelf} />

				</div>
			</div>
		);
	}

}


export default BookShelf;
