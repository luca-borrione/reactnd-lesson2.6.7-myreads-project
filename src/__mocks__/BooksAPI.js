import * as data from './allBooks.json'
import { TShelfKey } from '../shared/types'

const DELAY = 10;

const getDelay = () => DELAY * (Math.floor(Math.random() * 10) + 1); // random between 1 - 10

export const MOCKED = true;

let allBooks;
export const __resetBooks = () => {
	allBooks = JSON.parse(JSON.stringify(data.books)); // cloning
};
__resetBooks();

const getBooksInShelves = () =>
	allBooks.filter( ({ shelf }) => shelf !== TShelfKey.NONE );


export const getAll = () => {
	const booksInShelves = getBooksInShelves();
	return new Promise( resolve => {
		const delay = getDelay();
		setTimeout(() => {
			resolve(booksInShelves);
		}, delay);
	});
};

export const get = bookId => {
	const book = allBooks.find( ({ id }) => id === bookId );
	return new Promise( resolve => {
		const delay = getDelay();
		setTimeout(() => {
			resolve(book);
		}, delay);
	});
};

const formatUpdatedBooks = (books) => {
	const fBooks = {};

	Object.keys(TShelfKey).forEach( key => {
		const shelfKey = TShelfKey[key];
		if (shelfKey !== TShelfKey.NONE) {
			fBooks[shelfKey] = books
							.filter( ({ shelf }) => shelf === shelfKey )
							.map( book => book.id );
		}
	});

	return fBooks
};

export const update = (book, shelf) => {
	const existingBook = allBooks.find( ({ id }) => id === book.id );
	existingBook.shelf = shelf;
	const booksInShelves = getBooksInShelves();
	return new Promise( resolve => {
		const delay = getDelay();
		setTimeout(() => {
			resolve(formatUpdatedBooks(booksInShelves));
		}, delay);
	});
};

export const search = (query, maxResults) => {
	const re = new RegExp(`^${query}`, 'i');
	const result = allBooks.filter( ({ title, authors }) => {
		return re.test(title);
	})
	.slice(0, maxResults);

	return new Promise( resolve => {
		const delay = getDelay();
		setTimeout(() => {
			resolve(result);
		}, delay);
	});
};

