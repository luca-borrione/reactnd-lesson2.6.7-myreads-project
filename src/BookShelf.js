import React from 'react';
import BooksGrid from './BooksGrid';
import PropTypes from 'prop-types';
import { TBook } from './types';
import { ERROR, SHELF_TITLE } from './Constants';

/**
 * @class BookShelf
 * @extends React.Component
 * @classdesc
 * Prints the shelf title and shows the list of all the books in the shelf
 * @hideconstructor
 */
class BookShelf extends React.Component {

	/**
	 * @property {Object} propTypes - Intended types passed to the component
	 * @property {TBook[]} propTypes.books - List of all the books in the shelf
	 * @property {function} propTypes.updateBookShelf
	 * 	Method of the [App]{@link App#updateBookShelf} component for updating the list of books
	 * 	in the shelves
	 * @static
	 */
	static propTypes = {
		books: PropTypes.arrayOf(
			PropTypes.shape(TBook).isRequired
		).isRequired,
		updateBookShelf: PropTypes.func.isRequired
	};


	/**
	 * @description
	 * - Prints the shelf's title
	 * - Shows the list of books in the shelf by rendering the [BooksGrid]{@link BooksGrid} component
	 * @returns {ReactElement}
	 * @throws {Error} if the books passed to the component are contained in multiple shelves an exception will be thrown
	 */
	render() {
		const { books, updateBookShelf } = this.props;

		const shelfKeys = [
			...new Set(books.map( book => book.shelf ))
		];

		if (shelfKeys.length > 1) {
			throw new Error(ERROR.BOOKS_IN_MULTI_SHELVES, shelfKeys);
		}
		const shelfTitle = SHELF_TITLE[shelfKeys[0]];

		return (
			<div className="bookshelf">
				<h2 className="bookshelf-title">{shelfTitle}</h2>
				<div className="bookshelf-books">

					<BooksGrid
						books={books}
						updateBookShelf={updateBookShelf} />

				</div>
			</div>
		);
	}

}


export default BookShelf;
