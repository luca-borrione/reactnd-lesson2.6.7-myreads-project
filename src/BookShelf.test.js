import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import BookShelf from './BookShelf';
import * as BooksAPI from './BooksAPI'; // mocked
import { TShelfKey } from './types';
import { ERROR } from './Constants';

describe('BookShelf', () => {

	beforeEach(() => {
		BooksAPI.__resetBooks();
	});


	let books = [];
	const updateBookShelf = () => {};


	it('renders without crashing', () => {
		const div = document.createElement('div');
		ReactDOM.render(
			<BookShelf
				books={books}
				updateBookShelf={updateBookShelf} />, div);
		ReactDOM.unmountComponentAtNode(div);
	});


	it('renders correctly', () => {
		const tree = renderer.create(
			<BookShelf
				books={books}
				updateBookShelf={updateBookShelf} />
		)
		.toJSON();

		expect(tree).toMatchSnapshot();
	});


	it('expects books containing the same shelf', async () => {
		const allBooks = await BooksAPI.getAll();

		books = allBooks.filter( ({ shelf }) =>
					shelf === TShelfKey.WANT_TO_READ);

		const fn = () => {
			shallow(
				<BookShelf
					books={books}
					updateBookShelf={updateBookShelf} />
			);
		};

		expect(fn).not.toThrowError();
	});

	it('throws an error if it receives books containing different shelves', async () => {
		const allBooks = await BooksAPI.getAll();

		books = allBooks
					.filter( ({ shelf }) =>
						shelf === TShelfKey.WANT_TO_READ || shelf === TShelfKey.CURRENTLY_READING );

		const fn = () => {
			shallow(
				<BookShelf
					books={books}
					updateBookShelf={updateBookShelf} />
			);
		};

		expect(fn).toThrowError(ERROR.BOOKS_IN_MULTI_SHELVES);
	});
});
