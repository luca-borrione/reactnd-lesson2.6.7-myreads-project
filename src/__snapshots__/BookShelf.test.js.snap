// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookShelf renders correctly 1`] = `
<div
  className="bookshelf"
>
  <h2
    className="bookshelf-title"
  />
  <div
    className="bookshelf-books"
  >
    <ol
      className="books-grid"
    />
  </div>
</div>
`;
